android.os.BadParcelableException in cgeo.geocaching.maps.routing

Reported as new crash with 2020.04.15-RC on Google Play.
Several devices running Android 9 or 10.

```
java.lang.RuntimeException: 
 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:3488)
 
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3635)
 
  at android.app.servertransaction.LaunchActivityItem.execute (LaunchActivityItem.java:83)
 
  at android.app.servertransaction.TransactionExecutor.executeCallbacks (TransactionExecutor.java:135)
 
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:95)
 
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2175)
 
  at android.os.Handler.dispatchMessage (Handler.java:107)
 
  at android.os.Looper.loop (Looper.java:237)
 
  at android.app.ActivityThread.main (ActivityThread.java:7857)
 
  at java.lang.reflect.Method.invoke (Native Method)
 
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1076)
Caused by: android.os.BadParcelableException: 
 
  at android.os.Parcel.readParcelableCreator (Parcel.java:3059)
 
  at android.os.Parcel.readParcelable (Parcel.java:2981)
 
  at android.os.Parcel.readValue (Parcel.java:2883)
 
  at android.os.Parcel.readListInternal (Parcel.java:3324)
 
  at android.os.Parcel.readArrayList (Parcel.java:2467)
 
  at cgeo.geocaching.maps.routing.Route.<init> (Route.java:261)
 
  at cgeo.geocaching.maps.routing.Route$1.createFromParcel (Route.java:250)
 
  at cgeo.geocaching.maps.routing.Route$1.createFromParcel (Route.java:246)
 
  at android.os.Parcel.readParcelable (Parcel.java:2990)
 
  at android.os.Parcel.readValue (Parcel.java:2883)
 
  at android.os.Parcel.readArrayMapInternal (Parcel.java:3261)
 
  at android.os.BaseBundle.initializeFromParcelLocked (BaseBundle.java:292)
 
  at android.os.BaseBundle.unparcel (BaseBundle.java:236)
 
  at android.os.BaseBundle.getBoolean (BaseBundle.java:899)
 
  at android.app.Activity.restoreHasCurrentPermissionRequest (Activity.java:8279)
 
  at android.app.Activity.performCreate (Activity.java:7951)
 
  at android.app.Activity.performCreate (Activity.java:7944)
 
  at android.app.Instrumentation.callActivityOnCreate (Instrumentation.java:1307)
 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:3463)
```

Another similar exception on an Android 8 device:

```
java.lang.RuntimeException: 
 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2825)
 
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2900)
 
  at android.app.ActivityThread.-wrap11 (Unknown Source)
 
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1601)
 
  at android.os.Handler.dispatchMessage (Handler.java:105)
 
  at android.os.Looper.loop (Looper.java:251)
 
  at android.app.ActivityThread.main (ActivityThread.java:6589)
 
  at java.lang.reflect.Method.invoke (Native Method)
 
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:240)
 
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:767)
Caused by: android.os.BadParcelableException: 
 
  at android.os.Parcel.readParcelableCreator (Parcel.java:2889)
 
  at android.os.Parcel.readParcelable (Parcel.java:2815)
 
  at android.os.Parcel.readValue (Parcel.java:2718)
 
  at android.os.Parcel.readListInternal (Parcel.java:3146)
 
  at android.os.Parcel.readArrayList (Parcel.java:2348)
 
  at cgeo.geocaching.maps.routing.Route.<init> (Route.java:261)
 
  at cgeo.geocaching.maps.routing.Route$1.createFromParcel (Route.java:250)
 
  at cgeo.geocaching.maps.routing.Route$1.createFromParcel (Route.java:246)
 
  at android.os.Parcel.readParcelable (Parcel.java:2824)
 
  at android.os.Parcel.readValue (Parcel.java:2718)
 
  at android.os.Parcel.readArrayMapInternal (Parcel.java:3085)
 
  at android.os.BaseBundle.unparcel (BaseBundle.java:257)
 
  at android.os.BaseBundle.getBoolean (BaseBundle.java:825)
 
  at android.app.Activity.restoreHasCurrentPermissionRequest (Activity.java:7223)
 
  at android.app.Activity.performCreate (Activity.java:6974)
 
  at android.app.Instrumentation.callActivityOnCreate (Instrumentation.java:1214)
 
  at android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2778)
```

--- a/main/src/cgeo/geocaching/maps/routing/Route.java
+++ b/main/src/cgeo/geocaching/maps/routing/Route.java
@@ -64,7 +64,7 @@ public class Route implements Parcelable {
 
         // Parcelable methods
 
-        public static final Creator<RouteSegment> CREATOR = new Creator<RouteSegment>() {
+        public static final Parcelable.Creator<RouteSegment> CREATOR = new Parcelable.Creator<RouteSegment>() {
 
             @Override
             public RouteSegment createFromParcel(final Parcel source) {
@@ -78,10 +78,10 @@ public class Route implements Parcelable {
 
         };
 
-        RouteSegment (final Parcel parcel) {
-            item = parcel.readParcelable(null);
-            point = parcel.readParcelable(null);
-            points = parcel.readArrayList(null);
+        private RouteSegment (final Parcel parcel) {
+            item = parcel.readParcelable(RouteItem.class.getClassLoader());
+            point = parcel.readParcelable(Geopoint.class.getClassLoader());
+            points = parcel.readArrayList(Geopoint.class.getClassLoader());
             distance = parcel.readFloat();
         }
 
@@ -257,8 +257,8 @@ public class Route implements Parcelable {
 
     };
 
-    Route (final Parcel parcel) {
-        segments = parcel.readArrayList(null);
+    private Route (final Parcel parcel) {
+        segments = parcel.readArrayList(RouteSegment.class.getClassLoader());
     }
 
     @Override
@@ -271,4 +271,3 @@ public class Route implements Parcelable {
         dest.writeList(segments);
     }
 }
-

--- a/main/src/cgeo/geocaching/maps/routing/RouteItem.java
+++ b/main/src/cgeo/geocaching/maps/routing/RouteItem.java
@@ -56,14 +56,10 @@ public class RouteItem implements Parcelable {
         this.id = id;
     }
 
-    public static final Parcelable.Creator<RouteItem> CREATOR =
-            new Parcelable.Creator<RouteItem>() {
+    public static final Parcelable.Creator<RouteItem> CREATOR = new Parcelable.Creator<RouteItem>() {
                 @Override
                 public RouteItem createFromParcel(final Parcel in) {
-                    final CoordinatesType type = CoordinatesType.values()[in.readInt()];
-                    final String geocode = in.readString();
-                    final int id = in.readInt();
-                    return new RouteItem(type, geocode, id);
+                    return new RouteItem(in);
                 }
 
                 @Override
@@ -71,6 +67,12 @@ public class RouteItem implements Parcelable {
                     return new RouteItem[size];
                 }
             };
+
+    private RouteItem (final Parcel parcel) {
+        type = CoordinatesType.values()[parcel.readInt()];
+        geocode = parcel.readString();
+        id = parcel.readInt();
+    }
 
     @Override
     public int describeContents() {

